#30 Occurrence Details

CodeGlitch opened the pull request:
Made some glue for the awesome widgets someone already had made for "Clicking on a marker shows the Occurrence Details". Also added the Flexible widget to some of the Text widgets, for smaller screens.
Couldn't compile on Android without changing build.graddle to 3.3.1 from 3.2.1.
I left the gradle change out to not break someone else work.

CodeGlitch committed:
Merge pull request #1 from vostpt/develop

Merge 07/11/2019

CodeGlitch committed:
#30 Occurrence Details

Made some glue for the awesome widgets someone already had made for "Clicking on a marker shows the Occurrence Details". Also added the Flexible widget to some of the Text widgets, for smaller screens.
Couldn't compile on Android without changing build.graddle to 3.3.1 from 3.2.1.
I left the gradle change out to not break someone else work.

CodeGlitch committed:
Build Gradle Upgrade

Vanethos commented on lib/presentation/ui/home/home_page.dart:
```diff
@@ -382,16 +392,143 @@ class _MapWidgetState extends State<MapWidget> {
         });
   }
+
+  bool openOccurrence = false;
```
Do you think this can be inside the BLoC? The BLoC would expose a Stream of <bool> that would show or hide the content.

This would make sure that, if we go to a detail page and for some reason this page is destroyed, when we go back it can read from the BLoC that the widget is visible

Vanethos commented on lib/presentation/ui/home/home_page.dart:
```diff
@@ -382,16 +392,143 @@ class _MapWidgetState extends State<MapWidget> {
+  bool openOccurrence = false;
+  OccurrenceModel selectedOccurrence;
+
+  Widget getOccurrenceFloatingActionButton() {
```
Do you think this can be extracted in another class?

If the previous suggestion is accepted, this class would then have the HomeBloc as an argument

Vanethos commented on lib/presentation/ui/home/home_page.dart:
```diff
@@ -382,16 +392,143 @@ class _MapWidgetState extends State<MapWidget> {
+    return Container();
+  }
+
+  Widget removeSelectedOccurrenceFloatingActionButton() {
```
The same as above, this could be one widget that depending on the open/closed status shows one widget or the other. OR just the same widget that changes the icon and the onPressed.

Vanethos commented on lib/presentation/ui/home/home_page.dart:
```diff
@@ -382,16 +392,143 @@ class _MapWidgetState extends State<MapWidget> {
+    return Container();
+  }
+
+  Widget getOccurrenceTypeWidgetType() {
```
This can be confused with getOccurrenceTypeWidget() since both have as an output a Widget. Maybe you should rename the method.

As with the previous comments, this can be extracted to a different widget in a different class as a Stateless Widget

Vanethos replied:
Additionally, please put this widget in the `utils/occurrence` folder and follow the convention of the other widgets (for example, having as the background the `OccurrenceWidgetBackground` widget.

Vanethos commented on lib/presentation/ui/home/home_page.dart:
```diff
@@ -382,16 +392,143 @@ class _MapWidgetState extends State<MapWidget> {
+        ));
+  }
+
+  Widget getOccurrenceTypeWidget() {
```
This method is called getOccurrenceTypeWidget() but it it either shows the occurrence type or the full details of the occurrences. Maybe it should be renamed.

Additionally, this can also be extracted to its own different class

Vanethos commented on lib/presentation/ui/utils/occurrence/occurrence_location_widget.dart:
```diff
@@ -115,11 +117,13 @@ class OccurrenceLocationWidget extends StatelessWidget {
-                      getFormattedDate(this._lastUpdated, FORMAT_DATE_FULL),
-                  textAlign: TextAlign.center,
-                  style: styleLastUpdated(),
+                child: Flexible(
```
Shouldn't Flexible have always as a parent a Row/Column or a Flex Widget?

miguelpruivo replied:
Not having the `Flexible` widget as a direct descendant of a `Row`, `Column` or any other `Flex` widget, should break the UI.

Vanethos commented on lib/presentation/ui/utils/occurrence/occurrence_on_site_help_widget.dart:
```diff
@@ -102,19 +102,21 @@ class OccurrenceOnSiteHelpWidget extends StatelessWidget {
-                      this._lastUpdated, Constants.FORMAT_DATE_FULL),
-              textAlign: TextAlign.center,
-              style: styleLastUpdated(),
+            child: Flexible(
```
The same as above, shoulnd't the Flexible widget be a child of Row/Column/Flex?

```
  /// Creates a widget that controls how a child of a [Row], [Column], or [Flex]
  /// flexes.
```

Vanethos commented on lib/presentation/ui/utils/occurrence/occurrence_time_widget.dart:
```diff
@@ -78,12 +78,14 @@ class OccurrenceTimeWidget extends StatelessWidget {
-                        this._lastUpdated, Constants.FORMAT_DATE_FULL),
-                textAlign: TextAlign.center,
-                style: styleLastUpdated(),
+              child: Flexible(
```
See above comments

Vanethos reviewed (changes requested):
Please see requested changes

miguelpruivo reviewed (changes requested):
@CodeGlitch address the @Vanethos change requests.

CodeGlitch committed:
Fix Flexible correct place in widget hierarchy

Moving Flexible to be right after row/column

Vanethos commented:
@CodeGlitch Please check that you have now conflicts in your file: `lib/presentation/ui/home/home_page.dart`

CodeGlitch committed:
Some code improvements over the original commimt

Splitted most functions into new Classes, and created a new file for occurrence_small_info.dart

CodeGlitch commented:
I didn't plan to, but i managed to do some extra work on the code (I will get back on studying tomorrow!). I never used bloc, so, be nice! x)
Hoping this improved what was sent before, but I'm expecting some requested changes!
Cheers! /offline

Vanethos commented on lib/presentation/ui/home/home_page.dart:
```diff
@@ -366,13 +375,52 @@ class _MapWidgetState extends State<MapWidget> {
+  }
+
+  bool isFullInfo = false;
+  String selectedOccurrenceId = "";
```
Isn't this information on the bloc with `selectedOccurrenceSink`?

Vanethos commented on lib/presentation/ui/home/home_page.dart:
```diff
@@ -382,16 +430,53 @@ class _MapWidgetState extends State<MapWidget> {
             ),
-            onPressed: () => print("clicked"),
+            onPressed: () {
+              selectedOccurrenceId = occurrence.id;
```
Why are you using here both set state and putting data into the bloc? could you explain the logic?

Vanethos commented on lib/presentation/ui/home/home_page.dart:
```diff
@@ -438,3 +530,104 @@ class _PermissionWidget extends StatelessWidget {
+        margin: EdgeInsets.only(bottom: 5, right: 5),
+        child: FloatingActionButton(
+          onPressed: () {
+            homebloc.openOccurrenceSink.add(!open);
```
So that I can understand, so this is a Stateless widget that always has the same bool value? And you're adding always that bool value to the sink?

So you create 2 of these, one for the open and one for the close state?

Vanethos commented on lib/presentation/ui/utils/occurrence/occurrence_small_info.dart:
```diff
@@ -0,0 +1,45 @@
+
+  @override
+  Widget build(BuildContext context) {
+    return Align(
```
If this is inside the same card, consider using the `OccurrenceWidgetBackground` for consinstency

Vanethos reviewed (changes requested):
There are some small questions and requested changes, but overal great work!

Vanethos commented:
Also @CodeGlitch I will only be able to approve the PR after you solve the Conflcit Issues

Vanethos commented:
I am having one issue @CodeGlitch:
- Sometimes when I'm clicking on the same marker (does not happen with all markers) it will select a different occurrence and if I click again it will show another one.

see https://drive.google.com/open?id=1L1M0bWhBPR8wb69YQpWKRQ5S0XqYxSVi for video evidence

Also:
- If you click outside the info widget (eg: over the map), is it possible to close the widget?
- Since we'r already on the map, you don't need to show the map widget (@OldMetalmind we're on the same page on this one, correct?)
- When the widget is open, you are not showing the occurrence type
- Please change the icon colors to white

CodeGlitch committed:
Switched Widget to use OccurrenceWidgetBackground

CodeGlitch committed:
Several Changes to previous commits

I figured what is doing the wrong selection on the marker tap, I removed the marker width/height (had 100), looks like it works better without it, but it might still problematic on a area with a lot of Markers, only solution at some point is to zoom in.

Removed the bool use to control if it is open or not.

 I can't figure out how to remove the one to control which marker is selected, could use some help to figure how I can use a stream on this one!

CodeGlitch committed:
Problem with Marker width/height

Changing Marker width/height to 25, instead of none, this issue should be looked into.
Looks like it has an effect on it's position on the map, moving all markers off position, any ideias why that happens? If it's set to 100, shows on the place if should be, if we reduce it it goes wrong. Any ideias why this happens?

miguelpruivo commented on lib/presentation/ui/utils/occurrence/occurrence_small_info.dart:
```diff
@@ -16,15 +15,14 @@ class OccurrenceSmallInfoWidget extends StatelessWidget {
           children: <Widget>[
             Container(
-              padding: EdgeInsets.only(bottom: 35),
+              margin: EdgeInsets.only(bottom: 30),
```
Shouldn't the custom margins names be used here as well?

Vanethos commented:
@CodeGlitch one small thing: if you fixed the issues that me and @miguelpruivo point out, please mark the issue as "solved", this way we'll know what is still missing! :D

CodeGlitch committed:
Changed widget to use custom margins names

CodeGlitch committed:
Update home_page.dart

CodeGlitch commented:
I created a new breanch, updated, and submitted a new PR, #109. This one should be closed because its obsolete.
